File: Rotate Image/JavaScript/RotateImage-solution-1.js
/**
 * Each value in n row c column is the new value in c row n column
 * So starting from the n working to 0 we can remove a value and add it to the end of
 * c row.
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
    // iterate over the matrix rows from the end
    for(let i = matrix.length - 1; i >= 0; i--){
        // iterate over the columns of a row starting from 0
        for(let t = 0; t < matrix.length; t++) {
            // remove the first item in row i
            let num = matrix[i].shift();
            // add the item to the row number that matches the column number
            matrix[t].push( num );
        }
    }
};

/**
 * 0 1 2
 * 3 4 5
 * 6 7 8
 * 
 * 0 1 2 6
 * 3 4 5
 * 7 8
 * 
 * 0 1 2 6
 * 3 4 5 7
 * 8
 * 
 * 
 * 
 * 
 */

File: Rotate Image/Java/RotateImage-solution-1.java

/**
 * If we reverse each row in the matrix
 * each value in r row, c column moves to a index after adding the remainder of the values c and r from the matrix length
 * 
 * */
class Solution {
    public void rotate(int[][] matrix) {
        int l = matrix.length-1;
        // reverse each row in the matrix
        for(int r = 0; r < matrix.length; r++){
            for(int i = 0; i < matrix.length/2; i++) {
                int swap = matrix[r][i];
                matrix[r][i] = matrix[r][l-i];
                matrix[r][l-i] = swap;
            }
        }
    
                
        for(int r = 0; r < matrix.length-1; r++){
            // move each value in the matrix
            // we only  need to itinerate up to the length - the row number
            for(int c = 0; c < l-r; c++ ){
                int d = l - c - r;               
                int swap = matrix[r][c];
                matrix[r][c] = matrix[r+d][c+d];
                matrix[r+d][c+d] = swap;
            }
            
        }
    }
                

}